raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -290,6 +304,10 @@ private void onBecomeFollowerOfElectedLeader(FollowerState state) {
         stateMachine.becomeFollower(state.epoch());
         electionTimer.reset(electionTimeoutMs);
         resetConnections();
+
+        // After becoming a follower, we need to complete pending fetches so that
+        // they can be resent to the leader without waiting for their expiration
+        purgatory.completeAll(null);
```

Comment:
Should we also complete all in purgatory when we transit from a leader to a voted follower, since we know we'd not be the new leader at least for the next voted epoch?

Reply:
Also inside `maybeAppendAsLeader` should we call `completeAll` as well?

Reply:
I debated this. I figured as long as the election is ongoing, there is no strong reason to return from fetches since there will be nowhere to retry until the election has finished.

Reply:
For the second comment, note that `maybeAppendAsLeader` calls `updateLeaderEndOffset` which then completes purgatory operations.